db/ops.go

```diff
@@ -26,26 +28,48 @@ func (l *Loader) Insert(tableName string, primaryKey string, data map[string]str
 		l.entries[tableName] = map[string]*Operation{}
 	}
 
-	if _, found := l.entries[tableName][primaryKey]; found {
+	if _, found := l.entries[tableName][uniqueID]; found {
 		return fmt.Errorf("attempting to insert in table %q a primary key %q, that is already scheduled for insertion, insert should only be called once for a given primary key", tableName, primaryKey)
 	}
 
 	if l.tracer.Enabled() {
-		l.logger.Debug("primary key entry never existed for table, adding insert operation", zap.String("primary_key", primaryKey), zap.String("table_name", tableName))
+		l.logger.Debug("primary key entry never existed for table, adding insert operation", zap.String("primary_key", uniqueID), zap.String("table_name", tableName))
 	}
 
-	// we need to make sure to add the primary key in the data so that it gets created
-	data[table.primaryColumn.name] = primaryKey
-	l.entries[tableName][primaryKey] = l.newInsertOperation(table, primaryKey, data)
+	// we need to make sure to add the primary key in the data so that
+	// it gets created
+	for _, primary := range l.tables[tableName].primaryColumn {
+		data[primary.name] = primaryKey[primary.name]
+	}
+	l.entries[tableName][uniqueID] = l.newInsertOperation(table, primaryKey, data)
 	l.entriesCount++
 	return nil
 }
 
+func createKeyValuePairs(m map[string]string) string {
+	return fmt.Sprint(m)
+}
```

Comment:
Hmmm that seems incorrect, first iteration order of `m` is not guaranteed. We just need a constant order for the lifespan of the process, but even that, I'm not sure if it's respected or not in Go. Also, we should probably have explicit controls over the ID construction.

We should definitely have here also an optimal case for `1`.  I've benchmarked some implementation and here the best one I could come with right now:

```suggestion
func createKeyValuePairs(m map[string]string) string {
	if len(m) == 1 {
		for _, v := range m {
			return v
		}
	}

	i := 0
	keys := make([]*string, len(m))
	for k := range m {
		keys[i] = &k
		i++
	}

	sort.Slice(keys, func(i, j int) bool {
		return *keys[i] < *keys[j]
	})

	values := make([]string, len(keys))
	for i, key := range keys {
		values[i] = m[*key]
	}

	return strings.Join(values, "/")
}
```

And here the benchmark I used:

```
func Benchmark_createKeyValuePairs(b *testing.B) {
	for _, count := range []int{1, 2, 3, 4, 5} {
		b.Run(fmt.Sprintf("count=%d", count), func(b *testing.B) {
			m := make(map[string]string, count)
			for i := 0; i < count; i++ {
				m[fmt.Sprintf("%d", i)] = fmt.Sprintf("%d", i)
			}

			b.ResetTimer()
			b.ReportAllocs()
			for i := 0; i < b.N; i++ {
				createKeyValuePairs(m)
			}
		})
	}
}
```

Reply:
This code is even faster and with less allocations than `return fmt.Sprint(m)`

Reply:
Thank you for this beautiful code. Just updated.

Reply:
I tested the code and it doesn't work on this part:

> keys := make([]*string, len(m))
> 	for k := range m {
> 		keys[i] = &k
> 		i++
> 	}
> 	sort.Slice(keys, func(i, j int) bool {
> 		return *keys[i] < *keys[j]
> 	})

The pointer is referencing to the same address. I had to use `[]string` instead of `[]*string`

Reply:
Oh yeah that's a quirk of Golang where `&k` actually get the address of the variable that is doing the iteration internally. Let's go with the `[]string`, that was a micro-optimization.

I'll check to have the correct way of doing later.

Reply:
Sorry about it, I indeed just checked that the code was compiling correctly.